// zoom-vue/src/store/index.js
import { createStore } from 'vuex'

const store = createStore({
    state () {
      return {
        isGridView: false,
        reactionMain: null,
        userActive: null,
        userPin: null,
        userShareScreen: null,
        activeSpeakerMap: {},
        cmdRaisedHandMap: {},
        cmdReactionMap: {},
      }
    },
    mutations: {
        setGridView (state, isGridView) {
          state.isGridView = isGridView
        },
        setUserActive(state, userActive){
          state.userActive = userActive
        },
        setUserPin(state, userPin){
          state.userPin = userPin
        },
        setReactionMain(state, reactionMain){
          state.reactionMain = reactionMain
        },
        setUserShareScreen(state, userShareScreen){
          state.userShareScreen = userShareScreen;
        },
        setActiveSpeakerMap(state, activeSpeakerMap) {
          state.activeSpeakerMap = activeSpeakerMap
        },
        setRaisedHandMap (state, cmdRaisedHandMap) {
            state.cmdRaisedHandMap = cmdRaisedHandMap
        },
        setReactionMap (state, cmdReactionMap) {
            state.cmdReactionMap = cmdReactionMap
        },
    }
  })

  export default store
